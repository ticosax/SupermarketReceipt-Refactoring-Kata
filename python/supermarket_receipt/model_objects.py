from dataclasses import dataclass
from enum import Enum


class ProductUnit(Enum):
    EACH = 1
    KILO = 2


@dataclass(frozen=True)
class Product:
    name: str
    unit: ProductUnit


@dataclass
class ProductQuantity:
    product: Product
    quantity: float


class SpecialOfferType(Enum):
    THREE_FOR_TWO = 1
    TEN_PERCENT_DISCOUNT = 2
    TWO_FOR_AMOUNT = 3
    FIVE_FOR_AMOUNT = 4


@dataclass
class Offer:
    offer_type: SpecialOfferType
    product: Product
    argument: float


@dataclass
class Discount:
    product: Product
    description: str
    discount_amount: float
