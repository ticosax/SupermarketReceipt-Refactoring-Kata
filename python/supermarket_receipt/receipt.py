from dataclasses import dataclass, field
from math import fsum

from supermarket_receipt.model_objects import Discount, Product


@dataclass
class ReceiptItem:
    product: Product
    quantity: float
    price: float
    total_price: float


@dataclass
class Receipt:
    items: list[ReceiptItem] = field(default_factory=list)
    discounts: list[Discount] = field(default_factory=list)

    def total_price(self):
        total = fsum(item.total_price for item in self.items)
        total += fsum(discount.discount_amount for discount in self.discounts)
        return total

    def add_product(self, product, quantity, price, total_price):
        self.items.append(ReceiptItem(product, quantity, price, total_price))

    def add_discount(self, discount):
        self.discounts.append(discount)
