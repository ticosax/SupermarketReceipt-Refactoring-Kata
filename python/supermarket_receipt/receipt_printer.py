import mimetypes
from functools import partial

from jinja2 import Environment, PackageLoader, select_autoescape

from supermarket_receipt.model_objects import Discount, ProductUnit
from supermarket_receipt.receipt import Receipt, ReceiptItem


def render_item_line(columns: int, item: ReceiptItem):
    line_price = f"{item.total_price/ 100:.2f}"
    blank_space = " " * (columns - len(line_price) - len(item.product.name))
    return f"{item.product.name}{blank_space}{line_price}"


def render_total_line(columns: int, receipt: Receipt):
    line_price = f"{receipt.total_price () / 100:.2f}"
    blank_space = " " * (columns - len(line_price) - 6)
    return f"Total:{blank_space}{line_price}"


def render_discount_line(columns: int, discount: Discount):
    lhs = f"{discount.description} ({discount.product.name})"
    line_discount_amount = f"{discount.discount_amount / 100:.2f}"
    blank_space = " " * (columns - len(line_discount_amount) - len(lhs))
    return f"{lhs}{blank_space}{line_discount_amount}"


def is_each(unit: ProductUnit) -> bool:
    return unit is ProductUnit.EACH


class ReceiptPrinter:
    def __init__(self, columns=40):
        self.columns = columns

    def print_receipt(self, receipt: Receipt, mimetype: str = "text/plain"):
        env = Environment(
            loader=PackageLoader("supermarket_receipt"), autoescape=select_autoescape()
        )
        env.filters["asitemline"] = partial(render_item_line, self.columns)
        env.filters["astotalline"] = partial(render_total_line, self.columns)
        env.filters["asdiscountline"] = partial(render_discount_line, self.columns)
        env.filters["is_each"] = is_each
        extension = mimetypes.guess_extension(mimetype)
        template = env.get_template(f"receipt{extension}")
        return template.render(receipt=receipt, columns=self.columns)
