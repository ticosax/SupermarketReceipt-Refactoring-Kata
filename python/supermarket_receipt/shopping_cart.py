import math
from dataclasses import dataclass, field

from supermarket_receipt.model_objects import (
    Discount,
    Offer,
    Product,
    ProductQuantity,
    SpecialOfferType,
)
from supermarket_receipt.receipt import Receipt


def _n_for_amount(
    x: int,
    product: Product,
    quantity: float,
    offer: Offer,
    unit_price: int,
    quantity_as_int: int,
) -> Discount:
    number_of_x = math.floor(quantity_as_int / x)
    total = offer.argument * number_of_x + quantity_as_int % x * unit_price
    discount_n = unit_price * quantity - total
    return Discount(product, f"{x} for {offer.argument}", -discount_n)


def _discount_builder(
    product: Product,
    quantity: float,
    offer: Offer,
    unit_price: int,
    quantity_as_int: int,
) -> Discount | None:
    match offer.offer_type:
        case SpecialOfferType.TWO_FOR_AMOUNT if quantity_as_int >= 2:
            return _n_for_amount(
                2, product, quantity, offer, unit_price, quantity_as_int
            )
        case SpecialOfferType.FIVE_FOR_AMOUNT if quantity_as_int >= 5:
            return _n_for_amount(
                5, product, quantity, offer, unit_price, quantity_as_int
            )
        case SpecialOfferType.THREE_FOR_TWO if quantity_as_int >= 3:
            x = 3
            number_of_x = math.floor(quantity_as_int / x)
            discount_amount = quantity * unit_price - (
                (number_of_x * 2 * unit_price) + quantity_as_int % x * unit_price
            )
            return Discount(product, "3 for 2", -discount_amount)
        case SpecialOfferType.TEN_PERCENT_DISCOUNT:
            return Discount(
                product,
                str(offer.argument) + "% off",
                -quantity * unit_price * offer.argument / 100.0,
            )
        case _:
            message = f"Kind of offer not handled {offer.offer_type}"
            raise NotImplementedError(message)


@dataclass
class ShoppingCart:
    items: list[ProductQuantity] = field(default_factory=list)
    product_quantities: dict[Product, list[float]] = field(default_factory=dict)

    def add_item(self, product: Product) -> None:
        self.add_item_quantity(product, 1.0)

    def add_item_quantity(self, product: Product, quantity: float) -> None:
        self.items.append(ProductQuantity(product, quantity))
        self.product_quantities.setdefault(product, []).append(quantity)

    def handle_offers(self, receipt: Receipt, offers: dict[Product, Offer], catalog):
        for (
            p,
            quantities,
        ) in self.product_quantities.items():
            try:
                offer = offers[p]
            except KeyError:
                continue
            unit_price = catalog.unit_price(p)
            quantity = math.fsum(quantities)
            quantity_as_int = int(quantity)
            if discount := _discount_builder(
                p, quantity, offer, unit_price, quantity_as_int
            ):
                receipt.add_discount(discount)
