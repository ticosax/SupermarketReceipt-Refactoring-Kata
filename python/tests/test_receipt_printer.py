from approvaltests import verify

from supermarket_receipt.receipt_printer import ReceiptPrinter


def test_receipt_printer_wo_discount_text(no_discount_receipt):
    printer = ReceiptPrinter()
    verify(printer.print_receipt(no_discount_receipt, mimetype="text/plain"))


def test_receipt_printer_with_discount_text(percent_discount_receipt):
    printer = ReceiptPrinter()
    verify(printer.print_receipt(percent_discount_receipt, mimetype="text/plain"))


def test_receipt_printer_with_bulk_discount_text(bulk_discount_receipt):
    printer = ReceiptPrinter()
    verify(printer.print_receipt(bulk_discount_receipt, mimetype="text/plain"))


def test_receipt_printer_wo_discount_html(no_discount_receipt):
    printer = ReceiptPrinter()
    verify(printer.print_receipt(no_discount_receipt, mimetype="text/html"))


def test_receipt_printer_with_discount_html(percent_discount_receipt):
    printer = ReceiptPrinter()
    verify(printer.print_receipt(percent_discount_receipt, mimetype="text/html"))


def test_receipt_printer_with_bulk_discount_html(bulk_discount_receipt):
    printer = ReceiptPrinter()
    verify(printer.print_receipt(bulk_discount_receipt, mimetype="text/html"))
