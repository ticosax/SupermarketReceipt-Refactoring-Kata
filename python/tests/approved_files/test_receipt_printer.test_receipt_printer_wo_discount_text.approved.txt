apples                              1.80
  0.90 * 2.000
baguette                            3.90
  1.30 * 3

Total:                              5.70
